gazebo_plugins/src/gazebo_ros_vacuum_gripper.cpp

```diff
@@ -0,0 +1,247 @@
+  srv1_ = rosnode_->advertiseService(aso1);
+  ros::AdvertiseServiceOptions aso2 =
+    ros::AdvertiseServiceOptions::create<std_srvs::Empty>(
+    "off", boost::bind(&GazeboRosVacuumGripper::OffServiceCallback,
+    this, _1, _2), ros::VoidPtr(), &queue_);
+  srv2_ = rosnode_->advertiseService(aso2);
+
+  // Custom Callback Queue
+  callback_queue_thread_ = boost::thread( boost::bind( &GazeboRosVacuumGripper::QueueThread,this ) );
+
+  // New Mechanism for Updating every World Cycle
+  // Listen to the update event. This event is broadcast every
+  // simulation iteration.
+  update_connection_ = event::Events::ConnectWorldUpdateBegin(
+      boost::bind(&GazeboRosVacuumGripper::UpdateChild, this));
+
+  ROS_INFO("Loaded gazebo_ros_vacuum_gripper");
+}
+
+bool GazeboRosVacuumGripper::OnServiceCallback(std_srvs::Empty::Request &req,
+                                     std_srvs::Empty::Response &res)
+{
+  if (status_) {
+    ROS_WARN("gazebo_ros_vacuum_gripper: already status is 'on'");
+  } else {
+    status_ = true;
+    ROS_INFO("gazebo_ros_vacuum_gripper: status: off -> on");
+  }
+  return true;
+}
+bool GazeboRosVacuumGripper::OffServiceCallback(std_srvs::Empty::Request &req,
+                                     std_srvs::Empty::Response &res)
+{
+  if (status_) {
+    status_ = false;
+    ROS_INFO("gazebo_ros_vacuum_gripper: status: on -> off");
+  } else {
+    ROS_WARN("gazebo_ros_vacuum_gripper: already status is 'off'");
+  }
+  return true;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Update the controller
+void GazeboRosVacuumGripper::UpdateChild()
+{
+  std_msgs::Bool grasping_msg;
+  grasping_msg.data = false;
+  if (!status_) {
+    pub_.publish(grasping_msg);
+    return;
+  }
+  // apply force
+  lock_.lock();
+  math::Pose parent_pose = link_->GetWorldPose();
+  physics::Model_V models = world_->GetModels();
+  for (size_t i = 0; i < models.size(); i++) {
+    if (models[i]->GetName() == link_->GetName() ||
+        models[i]->GetName() == parent_->GetName())
+    {
+      continue;
+    }
+    physics::Link_V links = models[i]->GetLinks();
+    for (size_t j = 0; j < links.size(); j++) {
+      math::Pose link_pose = links[j]->GetWorldPose();
+      math::Pose diff = parent_pose - link_pose;
+      double norm = diff.pos.GetLength();
+      if (norm < 0.05) {
+        links[j]->SetLinearAccel(link_->GetWorldLinearAccel());
+        links[j]->SetAngularAccel(link_->GetWorldAngularAccel());
+        links[j]->SetLinearVel(link_->GetWorldLinearVel());
+        links[j]->SetAngularVel(link_->GetWorldAngularVel());
+        double norm_force = 1 / norm;
+        if (norm < 0.01) {
+          // apply friction like force
+          // TODO(unknown): should apply friction actually
+          link_pose.Set(parent_pose.pos, link_pose.rot);
+          links[j]->SetWorldPose(link_pose);
+        }
+        if (norm_force > 20) {
+          norm_force = 20;  // max_force
```

Comment:
it would be nice to be able to specify the `norm_force` and `norm` parameters via the plugin xml tags